source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -47,6 +47,13 @@ bool cbsContainsU16(CBS& cbs, uint16_t n) {
   return false;
 }
 
+std::string lastCryptoError() {
+  char errbuf[256] = {0};
+
+  ERR_error_string_n(ERR_get_error(), errbuf, sizeof(errbuf));
+  return std::string(errbuf);
+}
+
```

Comment:
Does this have any coverage? Are there RELEASE_ASSERTS that are firing for actual configuration errors? Do those need to be converted to exceptions? I'm curious how you came about this change and if something else needs to be fixed?

Reply:
This is split out from #9275. Unfortunately, there's no test coverage for these and I'm not really sure that they are testable. Mostly, I agree that the digest API uses deserve to be release asserts, but as I went through the code in #9275, I felt guilty about leaving a `RELEASE_ASSERT` without a message, so I added the crypto error string (if it's present) so that I could feel better about it.

I'm not addressing the asserts that ought to be exceptions yet, but plan to do that in a following PR. The next PR will return the hash rather than passing in a buffer that is used both as a temporary and as the output value, and the final PR will convert the assertion that fires for invalid certificates (no subject or alt names) into an exception.

Reply:
Well, then can we remove this until we have a test case that actually uses it and ends up throwing an exception? If you can't reproduce it with a test I don't think we should have it in the code.

Reply:
OK, I will update this to add a test (possibly death test?) that exercises `lastCryptoError`.

Reply:
I moved `lastCryptoError` into `tls/utility.cc` and added a test for it there.